assembly/src/assembler/module_graph/mod.rs

```diff
@@ -197,17 +322,31 @@ impl ModuleGraph {
             let module_id = ModuleIndex::new(high_water_mark + pending_index);
 
             // Apply module to call graph
-            for (index, procedure) in pending_module.procedures().enumerate() {
-                let procedure_id = ProcedureIndex::new(index);
-                let global_id = GlobalProcedureIndex {
-                    module: module_id,
-                    index: procedure_id,
-                };
-
-                // Ensure all entrypoints and exported symbols are represented in the call graph,
-                // even if they have no edges, we need them in the graph for the topological sort
-                if matches!(procedure, Export::Procedure(_)) {
-                    self.callgraph.get_or_insert_node(global_id);
+            match pending_module {
+                PendingModuleWrapper::Ast(pending_module) => {
```

Comment:
If compiled modules are added to the module graph "up-font" (i.e., at construction time), the probably won't make it into the pending module set and may not need to have special handling for compiled modules here.

Reply:
I think removing the pending set is intricately linked with #1416. The main issue right now is with module rewriting [here](https://github.com/0xPolygonMiden/miden-vm/blob/b8a767f3dc318db7b36502faec77dc6987bdeafe/assembly/src/assembler/module_graph/mod.rs#L228-L229), which is the only reason why we need pending modules to be `Box<Module>` instead of `Arc<Module>` (i.e. we need ownership to be able to mutate it). I am also assuming that using `Arc<Module>` in our public API is a requirement.

But all this module rewriting goes away if we only allow modules to be added all at once, and the graph to be "recomputed" only once (which can be done with #1416). And then removing the pending module list becomes easy.

Reply:
I think incremental additions of AST modules is something we may want to keep in the public API, but since we generate the output only once, we probably can do the "recompute" part only once as well. So, basically:

- All compiled modules are added at construction time.
- AST modules can be added one-by-one, but nothing is recomputed at this point.
- When the user calls `assemble_program()`, `assemble_library()` etc. the graph is computed and this happens only once.